internal/core/software_renderer.rs

```diff
@@ -1128,21 +1163,117 @@ impl<T: TargetPixel> RenderToBuffer<'_, T> {
         }
     }
 
+    fn foreach_region(
+        &mut self,
+        geometry: &PhysicalRect,
+        mut f: impl FnMut(&mut B, PhysicalRect, i16, i16),
+    ) {
+        let mut line = geometry.min_y();
+        while let Some(mut next) =
+            region_line_ranges(&self.dirty_region, line, &mut self.dirty_range_cache)
+        {
+            next = next.min(geometry.max_y());
+            for r in &self.dirty_range_cache {
+                if geometry.origin.x >= r.end {
+                    continue;
+                }
+                let begin = r.start.max(geometry.origin.x);
+                let end = r.end.min(geometry.origin.x + geometry.size.width);
+                if begin >= end {
+                    continue;
+                }
+                let extra_left_clip = begin - geometry.origin.x;
+                let extra_right_clip = geometry.origin.x + geometry.size.width - end;
+
+                let region = PhysicalRect {
+                    origin: PhysicalPoint::new(begin, line),
+                    size: PhysicalSize::new(end - begin, next - line),
+                };
+
+                f(&mut self.buffer, region, extra_left_clip, extra_right_clip);
+            }
+            if next == geometry.max_y() {
+                break;
+            }
+            line = next;
+        }
+    }
+
     fn process_texture_impl(&mut self, geometry: PhysicalRect, texture: SceneTexture<'_>) {
-        self.foreach_ranges(&geometry, |line, buffer, extra_left_clip, extra_right_clip| {
-            draw_functions::draw_texture_line(
-                &geometry,
-                PhysicalLength::new(line),
-                &texture,
-                buffer,
-                extra_left_clip,
-                extra_right_clip,
-            );
-        });
+        if !self.buffer.draw_texture(
```

Comment:
That was better before. If you are not passing the dirty region to the function, how is it going to clip for the dirty region?
So I see two possibility:
 1. Put pack the call to draw_texture within foreach_region
 2. Or pass the dirty region to the draw_texture function and let the implementer do the clipping.

I'd think 1. is the most convenient, unless there is a way to do hardware clip, but i don't think that's the case on most hardware i know.

Reply:
👍 I reverted back to 1.